### PR TITLE
Add AMP support for Pytorch

### LINKED ISSUE #471: Extend ProjectedGradientDescent for PyTorch with option for apex
Extend evasion attack `ProjectedGradientDescent` with option to use `apex` to enable mixed precision and distributed adversarial training in Pytorch.

- https://github.com/NVIDIA/apex

### PR DESCRIPTION
**Pull request by minhitbk**
# Description
This PR implement the amp support for pytorch classifier and therefore support all attacks that uses pytorch classifier, including PGD. In addition, a number of fixes are also done for asr attack and deepspeech estimator.

Fixes #471 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [x] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/estimators/classification/pytorch.py [hunk 1]
```diff
@@ -63,6 +63,9 @@ def __init__(
         input_shape: Tuple[int, ...],
         nb_classes: int,
         optimizer: Optional["torch.optim.Optimizer"] = None,  # type: ignore
+        use_amp: bool = False,
+        opt_level: str = "O1",
+        loss_scale: Optional[Union[float, str]] = "dynamic",
```
Comment (hesseltuinhof) [inline art/estimators/classification/pytorch.py:68]: In the ASR estimator and attack, a value of `1.0` is used. I initially didn't understand why strings were allowed. Maybe you could add the possible string arguments to the docstring?

**Review by hesseltuinhof — APPROVED** [review]
Besides one small suggestion, everything looks good! :+1:

File: art/estimators/classification/pytorch.py (hunk 1 above)
Comment (minhitbk) [reply art/estimators/classification/pytorch.py:68]: Thank you @hesseltuinhof , this fixed.

File: conftest.py [hunk 2]
```diff
@@ -764,3 +764,33 @@ def framework_agnostic(request, framework):
     if request.node.get_closest_marker("framework_agnostic"):
         if framework != default_framework:
             pytest.skip("framework agnostic test skipped for framework : {}".format(framework))
+
+
+@pytest.fixture
+def expected_values_amp(request):
+    """
```
Comment (killianlevacher) [inline conftest.py:771]: Is there any specific reason another expected_value fixture was created? Normally the existing `expected_values` fixture should work with any test. If for some reason, it does not work for this test, I think it would be better to alter the existing fixture as opposed to creating a new one. I can help out if the usage of the original `expected_values` fixture is unclear

File: tests/estimators/classification/test_deeplearning_common.py [hunk 3]
```diff
@@ -509,3 +509,94 @@ def test_learning_phase(art_warning, image_dl_estimator):
...
+apex_spec = importlib.util.find_spec("apex")
+if apex_spec is not None:
+    amp_spec = importlib.util.find_spec("apex.amp")
+else:
+    amp_spec = None
+amp_found = amp_spec is not None
+
+
+@pytest.mark.skipif(not amp_found, reason="Skip unittests if apex module is not found.")
+@pytest.mark.only_with_platform("pytorch")
```
Comment (killianlevacher) [inline tests/estimators/classification/test_deeplearning_common.py:524]: Will this test a) only ever be able to run on Pytorch only due to some Pytorch dependency, or b) is it that it only passes for Pytorch for now but could eventually pass for other frameworks? 
If case a): then `@pytest.mark.skipMlFramework("tensorflow", "keras", etc..... "non_dl_frameworks")` should be used instead of `@pytest.mark.only_with_platform("pytorch")` (`only_with_platform` is only currently used temporarily while the tests are being refactored)
if case b) then this test should be placed within `test_deeplearning_specific.py` instead of `test_deeplearning_common.py`

File: tests/estimators/classification/test_deeplearning_common.py [hunk 4]
```diff
@@ -509,3 +509,94 @@ def test_learning_phase(art_warning, image_dl_estimator):
         assert hasattr(classifier, "_learning_phase")
     except ARTTestException as e:
         art_warning(e)
+
+
+# Note, the following part is for testing the amp tool on pytorch platform only.
+apex_spec = importlib.util.find_spec("apex")
```
Comment (killianlevacher) [inline tests/estimators/classification/test_deeplearning_common.py:515]: just out of curiosity, do you think this `amp_found` check will need to be done in other testing scripts? If yes then it would make more sense to encapsulate this code as a new pytest marker, similar in spirit to @pytest.mark.skipMlFramework defined in `conftest.py/skip_by_platform`. This is really easy to do, I can help out if it's unclear how to encapsulate it

File: tests/estimators/classification/test_deeplearning_common.py [hunk 5]
```diff
@@ -509,3 +509,94 @@ def test_learning_phase(art_warning, image_dl_estimator):
...
+@pytest.mark.skipif(not amp_found, reason="Skip unittests if apex module is not found.")
+@pytest.mark.only_with_platform("pytorch")
+@pytest.mark.parametrize("device_type", ["cpu", "gpu"])
+def test_loss_gradient_amp(
+    get_default_mnist_subset,
+    image_dl_estimator,
+    expected_values_amp,
+    mnist_shape,
+    device_type,
+):
```
Comment (killianlevacher) [inline tests/estimators/classification/test_deeplearning_common.py:532]: this test is missing the default `try / except ARTTestException as e: art_warning(e)` at the beginning and end of the test. This is a safeguard which guarantees any ART Meta test Warnings are brought to our attention even if the test passes

File: tests/estimators/speech_recognition/test_pytorch_deep_speech.py [hunk 6]
```diff
@@ -121,6 +121,8 @@ def test_all(self, _test_all):
     @pytest.fixture(params=[False, True])
     def _test_all(self, request, setup_class):
         # Only import if deep speech module is available
+        import torch
```
Comment (killianlevacher) [inline tests/estimators/speech_recognition/test_pytorch_deep_speech.py:124]: this test is missing the default `try / except ARTTestException as e: art_warning(e)` same as above

File: tests/estimators/speech_recognition/test_pytorch_deep_speech.py [hunk 7]
```diff
@@ -368,6 +374,10 @@ def _test_all(self, request, setup_class):
             assert not ((transcriptions1 == transcriptions2).all())
 
         else:
+            # Create the optimizer
+            parameters = self.speech_recognizer.model.parameters()
+            self.speech_recognizer._optimizer = torch.optim.SGD(parameters, lr=0.01)
+
             # Before train
             transcriptions1 = self.speech_recognizer.predict(self.x, batch_size=2, transcription_output=True)
```
Comment (killianlevacher) [inline tests/estimators/speech_recognition/test_pytorch_deep_speech.py:382]: lines  (361 to 366) and (377 to 382) are duplicates they should be moved to before line 360

File: tests/estimators/speech_recognition/test_pytorch_deep_speech.py (hunk 6 above)
Comment (killianlevacher) [inline tests/estimators/speech_recognition/test_pytorch_deep_speech.py:124]: fixtures shouldn't be used in this way. This test code should simply be placed within the `def test_all(self, _test_all):` function directly. And instead use the `@pytest.mark.parametrize` marker to set a given param to various values see `test_loss_functions` in  `https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/dev_1.5.0/tests/estimators/classification/test_deeplearning_common.py`

File: tests/estimators/speech_recognition/test_pytorch_deep_speech.py [hunk 8]
```diff
@@ -356,6 +358,10 @@ def _test_all(self, request, setup_class):
 
         # Now test fit function
         if request.param is True:
+            # Create the optimizer
```
Comment (killianlevacher) [inline tests/estimators/speech_recognition/test_pytorch_deep_speech.py:361]: In general there is a few simple but important changes that need to be performed on this test to conform to the rest of ART standard tests. 1/ tests should not be placed within a class `class TestPyTorchDeepSpeech` 2/ the `master_seed(seed=1234)` needs to be remove as it is set in 1 location only for the entire set of tests 3/ all of the large np.asarrays should be stored and retrieved using the `store_expected_values` and `expected_values` ART fixtures to increase visibility. I can help out on how to use them if necessary, It's really easy

File: conftest.py (hunk 2 above)
Comment (minhitbk) [reply conftest.py]: Solved.

File: tests/estimators/classification/test_deeplearning_common.py (hunk 3 above)
Comment (minhitbk) [reply tests/estimators/classification/test_deeplearning_common.py]: Solved.

File: tests/estimators/classification/test_deeplearning_common.py (hunk 4 above)
Comment (minhitbk) [reply tests/estimators/classification/test_deeplearning_common.py]: Solved.

File: tests/estimators/classification/test_deeplearning_common.py (hunk 5 above)
Comment (minhitbk) [reply tests/estimators/classification/test_deeplearning_common.py:523]: Solved.

File: tests/estimators/speech_recognition/test_pytorch_deep_speech.py (hunk 7 above)
Comment (minhitbk) [reply tests/estimators/speech_recognition/test_pytorch_deep_speech.py]: Solved.

File: tests/estimators/speech_recognition/test_pytorch_deep_speech.py (hunk 6 above)
Comment (minhitbk) [reply tests/estimators/speech_recognition/test_pytorch_deep_speech.py]: Solved.

File: tests/estimators/speech_recognition/test_pytorch_deep_speech.py (hunk 8 above)
Comment (minhitbk) [reply tests/estimators/speech_recognition/test_pytorch_deep_speech.py]: Solved.

**Comment by minhitbk** [comment]
@killianlevacher : Thanks for the reviews on the test. I have updated the tests to reflect your reviews and followed the instructions on ART unittests style.

File: tests/estimators/classification/test_deeplearning_common.py [hunk 9]
```diff
@@ -509,3 +508,90 @@ def test_learning_phase(art_warning, image_dl_estimator):
         assert hasattr(classifier, "_learning_phase")
     except ARTTestException as e:
         art_warning(e)
+
+
+@pytest.mark.skipModule("apex.amp")
```
Comment (killianlevacher) [inline tests/estimators/classification/test_deeplearning_common.py:513]: Hi Minh, I'm just realising now this minor issue. The previous message might have been confusing. Just to be clear, `def test_loss_gradient_amp` should be placed in `test_deeplearning_specific.py` since it is specific to pytorch. Apart from that all the tests look good ! Thanks for upgrading them to the new system :-)

Comment (minhitbk) [reply tests/estimators/classification/test_deeplearning_common.py]: @killianlevacher : Done.

**Comment by beat-buesser** [comment]
@minhitbk @killianlevacher @hesseltuinhof Thank you all very much for your work on this great pull request!
